lib/can/src/CANInterface.cpp

```diff
@@ -0,0 +1,157 @@
+  m_rxEventQueue.call(callback(this, &CANInterface::rxPostman));
+}
+
+void CANInterface::rxPostman(void) {
+  CANMsg msg;
+  // this loop is needed to avoid missing msg received between turning off the IRQ and turning it back on
+  while (m_activeCANBus->read(msg)) {
+    CANMsg *mail = m_rxMailbox.try_alloc();  // no wait
+
+    if (mail != nullptr) {
+      *mail = msg;
+      MBED_ASSERT(m_rxMailbox.put(mail) == osOK);
+    } else {
+      MBED_WARNING(MBED_MAKE_ERROR(MBED_MODULE_PLATFORM, MBED_ERROR_CODE_BUFFER_FULL), "CAN RX mailbox full");
+    }
+  }
+  can_irq_set(m_activeCANBus->getHandle(), IRQ_RX, true);
+}
+
+void CANInterface::rxClient(void) {
+  while (true) {
+    CANMsg *mail = nullptr;
+
+    do {
+      mail = m_rxMailbox.try_get();  // TODO: try_get_for was not working. Investigate why and use it
+      ThisThread::sleep_for(1ms);
+    } while (mail == nullptr);
+
+    // Extract message
+    CANMsg msg = *mail;
+    MBED_ASSERT(m_rxMailbox.free(mail) == osOK);
+
+    // If message is one-shot, process message
+    if (m_rxOneShotMsgHandler->find(msg.getID()) != m_rxOneShotMsgHandler->end()) {
+      m_rxOneShotMsgHandler->at(msg.getID())(msg);
+    }
+
+    // Otherwise message is streamed - extract message signals and put in RX message map
+    else {
+      HWBRIDGE::CANMsgData_t msgData;
+      msg.getPayload(msgData);
+
+      if (!HWBRIDGE::unpackCANMsg(msgData.raw, msg.getID(), m_rxStreamedMsgMap)) {
+        MBED_WARNING(MBED_MAKE_ERROR(MBED_MODULE_PLATFORM, MBED_ERROR_CODE_INVALID_DATA_DETECTED),
+                     "CAN RX message unpacking failed");
+      }
+    }
+  }
+}
+
+void CANInterface::txProcessor(void) {
+  while (true) {
+    auto startTime = Kernel::Clock::now();
+
+    CANMsg *mail = nullptr;
+
+    // Send all one-shots messages that were queued
+    while ((mail = m_txMailboxOneShot.try_get()) != nullptr) {
+      if (!m_activeCANBus->write(*mail)) {
+        MBED_WARNING(MBED_MAKE_ERROR(MBED_MODULE_PLATFORM, MBED_ERROR_CODE_WRITE_FAILED), "CAN TX write failed");
+      }
+      MBED_ASSERT(m_txMailboxOneShot.free(mail) == osOK);
+      ThisThread::sleep_for(TX_INTERDELAY);
+    }
+
+    // Send all streamed messages
+    for (auto it = m_txStreamedMsgMap->begin(); it != m_txStreamedMsgMap->end(); it++) {
+      HWBRIDGE::CANID msgID = it->first;
+
+      HWBRIDGE::CANMsgData_t msgData = {0};
+      size_t len                     = 0;
+      if (HWBRIDGE::packCANMsg(msgData.raw, msgID, m_txStreamedMsgMap, len)) {
+        // Send message
+        CANMsg msg;
+        msg.setID(msgID);
+        msg.setPayload(msgData, len);
+        m_activeCANBus->write(msg);
+        ThisThread::sleep_for(TX_INTERDELAY);
+      } else {
+        MBED_WARNING(MBED_MAKE_ERROR(MBED_MODULE_PLATFORM, MBED_ERROR_CODE_INVALID_DATA_DETECTED),
+                     "CAN RX message packing failed");
+      }
+    }
+
+    ThisThread::sleep_until(startTime + TX_PERIOD);
```

Comment:
Replaced `ThisThread::sleep_for()` with `ThisThread::sleep_until()` cuz the TX interdelay is already 1ms, so TX messages weren't actually getting sent every 10ms. Although, if there's more than 10 TX messages to send then the TX interdelay will still make the transmit period greater than 10ms.